tests/testthat/test-epichains.R

```diff
@@ -0,0 +1,206 @@
+)
+
+test_that("print.epichains works for simulate_summary output", {
+  expect_snapshot(epichains_summary)
+})
+
+test_that("print.epichains works for simulate_tree output", {
+  expect_snapshot(epichains_tree)
+})
+
+test_that("print.epichains works for simulate_tree output", {
+  expect_snapshot(epichains_tree2)
+})
+
+test_that("summary.epichains works as expected", {
+  expect_named(
+    summary(epichains_summary),
+    c(
+      "chain_ran",
+      "max_chain_stat",
+      "min_chain_stat"
+    )
+  )
+  expect_named(
+    summary(epichains_tree2),
+    c(
+      "chains_ran",
+      "max_time",
+      "unique_ancestors",
+      "max_generation"
+    )
+  )
+  expect_named(
+    summary(epichains_tree2),
+    c(
+      "chains_ran",
+      "max_time",
+      "unique_ancestors",
+      "max_generation"
+    )
+  )
+  expect_true(
+    is.infinite(
+      summary(epichains_summary_all_infs)$min_chain_stat
+    )
+  )
+  expect_true(
+    is.infinite(
+      summary(epichains_summary_all_infs)$max_chain_stat
+    )
+  )
+})
+
+test_that("validate_epichains works", {
+  expect_invisible(
+    validate_epichains(epichains_summary)
+  )
+  expect_invisible(
+    validate_epichains(epichains_tree)
+  )
+  expect_invisible(
+    validate_epichains(epichains_tree2)
+  )
+})
+
+test_that("is_chains_tree works", {
+  expect_true(
+    is_chains_tree(epichains_tree)
+  )
+  expect_true(
+    is_chains_tree(epichains_tree2)
+  )
+  expect_false(
+    is_chains_tree(epichains_summary)
+  )
+})
+
+test_that("is_chains_summary works", {
+  expect_true(
+    is_chains_tree(epichains_tree)
+  )
+  expect_true(
+    is_chains_tree(epichains_tree2)
+  )
+  expect_false(
+    is_chains_tree(epichains_summary)
+  )
+})
+
+test_that("is_epichains_aggregate_df works", {
+  expect_true(
+    is_epichains_aggregate_df(aggreg_by_gen)
+  )
+  expect_true(
+    is_epichains_aggregate_df(aggreg_by_time)
+  )
+  expect_true(
+    is_epichains_aggregate_df(aggreg_by_both)
+  )
+  expect_false(
+    is_epichains_aggregate_df(epichains_tree)
+  )
+})
+
+test_that("validate_epichains throws errors", {
+  expect_error(
+    validate_epichains(mtcars),
+    "must have an epichains class"
+  )
+})
+
+test_that("head and tail methods work", {
+  expect_snapshot(head(epichains_tree))
+  expect_snapshot(head(epichains_tree2))
+  expect_snapshot(tail(epichains_tree))
+  expect_snapshot(tail(epichains_tree2))
+})
+
+test_that("aggregate method work", {
+  expect_named(
+    aggreg_by_gen,
+    c("generation", "cases")
+  )
+  expect_named(
+    aggreg_by_time,
+    c("time", "cases")
+  )
+  expect_identical(
+    as.vector(
+      vapply(aggreg_by_both, names, FUN.VALUE = character(2))
+    ),
+    c("time", "cases", "generation", "cases")
+  )
+  expect_s3_class(
+    aggreg_by_gen,
+    "epichains_aggregate_df"
+  )
+  expect_s3_class(
+    aggreg_by_time,
+    "epichains_aggregate_df"
+  )
+  expect_s3_class(
+    aggreg_by_both,
+    "epichains_aggregate_df"
+  )
```

Comment:
Here it might be worth also checking that the data aggregated by "both" inherits from a list, whereas the other two inherit from data.frame. I think there should be a wider rethinking of classes in {epichains} as well to avoid this ambiguous inheritance structure

Reply:
Thanks for the suggestion for rethink the class structures. I'll raise an issue for further discussion. For now, I've added the suggested test here fdf45f8e37304fe1fa7b323903227b51ce09c87d.